### PR TITLE
Added Kosaraju's Algorithm

### LINKED ISSUE #114: Content: Strongly Connected Components - Kosaraju's Algorithm
## Description

Add the article for Strongly Connected Components - Kosaraju's Algorithm. Use the following structure: 
- Introduction
- Explanation
- Algorithm
- Code
- Sample I/O
- Complexity Analysis

### PR DESCRIPTION
**Pull request by AshuKV**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

(Write your answer here.)

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Fixes #114

### CODE REVIEW

File: docs/graph-algorithms/kosarajus-algorithm.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+3. For each element u of L in order, do Assign(u,u) where Assign(u,root) is the recursive subroutine:
+    - If u has not been assigned to a component then:
+      1. Assign u as belonging to the component whose root is root.
+      2. For each in-neighbour v of u, do Assign(v,root).
+
+
+## Code
+
+### cpp
```
Comment (HarshCasper) [inline docs/graph-algorithms/kosarajus-algorithm.md:75]: That's not how you should add Code. Check the existing articles to know how to do it

File: docs/graph-algorithms/kosarajus-algorithm.md [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  3. Go to the previous vertex and visit its child vertices. But, all of its child vertices are visited, so push it into the stack. 
+    [![graph-main.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-step-3.png)](https://postimg.cc/Q9nzFp23)
+
+  4. Similarly, a final stack is created. 
+    [![graph-main.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-step-4.png)](https://postimg.cc/Q9nzFp23)
+
+
+- Reverse the original graph.
+
+    [![graph-1.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-reversed-graph.png)](https://postimg.cc/yk2XHcsL)
```
Comment (HarshCasper) [inline docs/graph-algorithms/kosarajus-algorithm.md:40]: You should NEVER use images from other websites

File: docs/graph-algorithms/kosarajus-algorithm.md [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+- Perform DFS on the reversed graph, with the source vertex as the vertex on top of the stack. When DFS finishes, all nodes visited will form one Strongly Connected Component. If any more nodes remain unvisited, this means there are more Strongly Connected Component's, so pop vertices from top of the stack until a valid unvisited node is found. Repeat this step until all nodes are visited.
+
+## Quick Example : 
+
+- Perform a dfs on the whole graph. Let us start from vertex-0, visit all of its child vertices, and mark the visited vertices as done. If a vertex leads to an already visited vertex, then push this vertex to the stack. 
+
+  For example: 
+  1. Starting from vertex-0, go to vertex-1, vertex-2, and then to vertex-3. Vertex-3 leads to already visited vertex-0, so push the source vertex into the stack.
+
+    [![graph-main.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-step-1.png)](https://postimg.cc/Q9nzFp23)
```
Comment (HarshCasper) [inline docs/graph-algorithms/kosarajus-algorithm.md:25]: You should NEVER use images from other websites

File: docs/graph-algorithms/kosarajus-algorithm.md [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+  2. Go to the previous vertex and visit its child vertices i.e. vertex-4, vertex-5, vertex-6 and vertex-7 sequentially. Since there is nowhere to go from vertex-7, push it into the stack. 
+  
+    [![graph-main.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-step-2.png)](https://postimg.cc/Q9nzFp23)
```
Comment (HarshCasper) [inline docs/graph-algorithms/kosarajus-algorithm.md:29]: You should NEVER use images from other websites

File: docs/graph-algorithms/kosarajus-algorithm.md [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+  3. Go to the previous vertex and visit its child vertices. But, all of its child vertices are visited, so push it into the stack. 
+    [![graph-main.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-step-3.png)](https://postimg.cc/Q9nzFp23)
+
+  4. Similarly, a final stack is created. 
+    [![graph-main.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-step-4.png)](https://postimg.cc/Q9nzFp23)
```
Comment (HarshCasper) [inline docs/graph-algorithms/kosarajus-algorithm.md:35]: You should NEVER use images from other websites

File: docs/graph-algorithms/kosarajus-algorithm.md [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  1. Starting from vertex-0, go to vertex-1, vertex-2, and then to vertex-3. Vertex-3 leads to already visited vertex-0, so push the source vertex into the stack.
+
+    [![graph-main.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-step-1.png)](https://postimg.cc/Q9nzFp23)
+
+  2. Go to the previous vertex and visit its child vertices i.e. vertex-4, vertex-5, vertex-6 and vertex-7 sequentially. Since there is nowhere to go from vertex-7, push it into the stack. 
+  
+    [![graph-main.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-step-2.png)](https://postimg.cc/Q9nzFp23)
+
+  3. Go to the previous vertex and visit its child vertices. But, all of its child vertices are visited, so push it into the stack. 
+    [![graph-main.png](https://cdn.programiz.com/sites/tutorial2program/files/scc-step-3.png)](https://postimg.cc/Q9nzFp23)
```
Comment (HarshCasper) [inline docs/graph-algorithms/kosarajus-algorithm.md:32]: You should NEVER use images from other websites

File: docs/graph-algorithms/kosarajus-algorithm.md [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+  g.add_edge(1, 2)
+  g.add_edge(2, 3)
+  g.add_edge(2, 4)
+  g.add_edge(3, 0)
+  g.add_edge(4, 5)
+  g.add_edge(5, 6)
+  g.add_edge(6, 4)
+  g.add_edge(6, 7)
+  print("The SCCs are:")
+  g.print_scc()
```
Comment (HarshCasper) [inline docs/graph-algorithms/kosarajus-algorithm.md:247]: Use dynamic input

File: docs/graph-algorithms/kosarajus-algorithm.md [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    Graph g(8);
+    g.addEdge(0, 1);
+    g.addEdge(1, 2);
+    g.addEdge(2, 3);
+    g.addEdge(2, 4);
+    g.addEdge(3, 0);
+    g.addEdge(4, 5);
+    g.addEdge(5, 6);
+    g.addEdge(6, 4);
+    g.addEdge(6, 7);
```
Comment (HarshCasper) [inline docs/graph-algorithms/kosarajus-algorithm.md:172]: Use dynamic input

File: docs/graph-algorithms/kosarajus-algorithm.md [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+2 - 5 : 2
+
+2 - 3 : 3
+
+3 - 4 : 3
+
+0 - 1 : 4
+```
+
+## Complexity Analysis
```
Comment (HarshCasper) [inline docs/graph-algorithms/kosarajus-algorithm.md:295]: Give a blank line after heading

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: docs/graph-algorithms/kosarajus-algorithm.md (hunk 1 above)
Comment (AshuKV) [reply docs/graph-algorithms/kosarajus-algorithm.md]: Updated.

**Comment by AshuKV** [comment]
Hi @HarshCasper, Please do review this PR and merge it, so that reference can be taken from here for related issues.

**Comment by AshuKV** [comment]
> Hi @HarshCasper, Please do review this PR and merge it, so that reference can be taken from here for related issues.

Any update pls Sir?

**Review by ankitaggarwal23 — APPROVED** [review]

**Pull request merged**